stylus-sdk/src/storage/bytes.rs

```diff
@@ -65,121 +55,117 @@ impl StorageBytes {
+            // if representation hasn't changed, just update the length
+            root.write_len(len);
+        } else if (len < 32) && (old > 32) {
+            // if shrinking, pull data in
+            root.word = Storage::get_word(*root.storage.base());
+            root.write_len(len);
         } else {
-            // place the len in the root with the long bit high
-            Storage::set_word(self.root, U256::from(len * 2 + 1).into())
+            // if growing, push data out
+            root.word[31] = 0; // clear len byte
+            Storage::set_word(*root.storage.base(), root.word);
+            root.write_len(len);
         }
     }
 
     /// Adds a byte to the end.
-    pub fn push(&mut self, b: u8) {
-        let index = self.len();
-        let value = U8::from(b);
-
-        macro_rules! assign {
-            ($slot:expr) => {
+    pub fn push(&mut self, value: u8) {
+        let mut root = BytesRoot::new_mut(self);
+        let index = root.len();
+
+        match index {
+            // still short representation after adding a byte
+            // add the byte and update length
+            ..31 => {
+                root.word[index] = value;
                 unsafe {
-                    Storage::set_uint($slot, index % 32, value); // pack value
-                    self.write_len(index + 1);
+                    root.write_len(index + 1);
                 }
-            };
-        }
-
-        if index < 31 {
-            return assign!(self.root);
-        }
-
-        // convert to multi-word representation
-        if index == 31 {
+            }
+            // convert to multi-word representation
             // copy content over (len byte will be overwritten)
-            let word = Storage::get_word(self.root);
-            unsafe { Storage::set_word(*self.base(), word) };
+            31 => {
+                root.word[index] = value;
+                unsafe {
+                    Storage::set_word(*root.storage.base(), root.word);
+                    root.write_len(index + 1);
+                }
+            }
+            // already long representation
+            // add the new byte and update length
+            _ => {
+                let slot = root.storage.base() + U256::from(index / 32);
+                unsafe {
+                    Storage::set_uint(slot, index % 32, U8::from(value));
+                    root.write_len(index + 1);
+                }
+            }
         }
-
-        let slot = self.base() + U256::from(index / 32);
-        assign!(slot);
     }
 
     /// Removes and returns the last byte, if it exists.
     /// As an optimization, underlying storage slots are only erased when all bytes in
     /// a given word are freed when in the multi-word representation.
     pub fn pop(&mut self) -> Option<u8> {
-        let len = self.len();
+        let mut root = BytesRoot::new_mut(self);
+        let len = root.len();
+        let index = len - 1;
```

Comment:
This will underflow if len equals 0, which may lead to panic in debug mode (I think).
You can move this variable to after the check below.